project/src/main/java/com/google/autograder/servlets/helpers/Client.java

```diff
@@ -1,48 +1,59 @@
 package com.google.autograder.servlets.helpers;
 
 import java.io.File;
+import java.util.Optional;
+import java.io.InputStream;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Paths;
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
 import org.json.simple.JSONObject;
+import java.util.stream.Collectors;
 import org.json.simple.parser.JSONParser;
+import java.nio.charset.StandardCharsets;
+import org.json.simple.parser.ParseException;
 
 public final class Client {
 
-    public static final String CLIENT_ID = getClientID();
-    public static final String CLIENT_SECRET = getClientSecret();
+    private static final String CREDENTIALS_PATH = "auth/credentials.json";
+    private static final JSONObject CREDENTIALS_WEB_JSON_CONTENT = retrieveCredentialsWebContent().get();
 
-    private static final String RESOURCES_PATH = "../../../project/src/main/resources"; // "../classes"
-    private static final String CREDENTIALS_PATH = "/auth/credentials.json";
+    public static final String CLIENT_ID = retrieveClientID().get();
+    public static final String CLIENT_SECRET = retrieveClientSecret().get();
 
-    private static String getClientID() {
-        String client_id = null;
+    private static Optional<JSONObject> retrieveCredentialsWebContent() {
```

Comment:
Since this is private, you can just have this return JSONObject and throw an exception from here if it's unable to find the credentials. Typically you'd also want constants to be immutable, which JSONObject I believe isn't. You can simplify this by doing something like:
```
public static final String CLIENT_ID;
public static final String CLIENT_SECRET;

static {
  String fileContents = ...;
  JSONObject json;
  try {
    json = new JSONParser.parse(fileContents);
  } catch (ParseException e) {
    throw new RuntimeException(e); // Since ParseException is a checked exception, and you can't throw checked exceptions from static initializer blocks
  }
  CLIENT_ID = json.get("web").get("client_id").toString();
  // similar for CLIENT_SECRET
}
```
This way you also don't depend on the order of the variables you declare (if you accidentally move line 20 below lines 22,23 then you'll get an error since lines 22,23 depend on line 20)

Reply:
Okay this makes a lot sense. Thank you.

Also +1 on re-throwing the ParseException as a RuntimeException, I didn't know you could do that.